rclcpp_action/test/test_server_goal_handle.cpp

```
@@ -0,0 +1,159 @@
+
+  bool cancel() {return try_canceling();}
+
+  void cancel_goal() {_cancel_goal();}
+};
+
+class TestServerGoalHandle : public ::testing::Test
+{
+public:
+  TestServerGoalHandle()
+  : handle_(nullptr) {}
+
+  void SetUp()
+  {
+    std::shared_ptr<rcl_action_goal_handle_t> rcl_handle =
+      std::make_shared<rcl_action_goal_handle_t>();
+    *rcl_handle.get() = rcl_action_get_zero_initialized_goal_handle();
+    rcutils_allocator_t allocator = rcutils_get_default_allocator();
+    rcl_action_goal_info_t goal_info = rcl_action_get_zero_initialized_goal_info();
+    ASSERT_EQ(RCL_RET_OK, rcl_action_goal_handle_init(rcl_handle.get(), &goal_info, allocator));
+    rclcpp_action::GoalUUID uuid;
+    std::shared_ptr<const test_msgs::action::Fibonacci::Goal> goal =
+      std::make_shared<const test_msgs::action::Fibonacci::Goal>();
+    auto on_terminal_state = [](const rclcpp_action::GoalUUID &, std::shared_ptr<void>) {};
+    auto on_executing = [](const rclcpp_action::GoalUUID &) {};
+    auto publish_feedback =
+      [](std::shared_ptr<test_msgs::action::Fibonacci::Impl::FeedbackMessage>) {};
+    handle_ = std::make_unique<FibonacciServerGoalHandle>(
+      rcl_handle, uuid, goal, on_terminal_state, on_executing, publish_feedback);
+  }
+
+protected:
+  std::unique_ptr<FibonacciServerGoalHandle> handle_;
+};
+
+TEST_F(TestServerGoalHandle, construct_destruct) {
+  EXPECT_FALSE(handle_->is_canceling());
+  EXPECT_TRUE(handle_->is_active());
+  EXPECT_FALSE(handle_->is_executing());
+}
+
+TEST_F(TestServerGoalHandle, cancel) {
+  handle_->execute();
+  EXPECT_TRUE(handle_->cancel());
+  EXPECT_FALSE(handle_->is_canceling());
+  EXPECT_FALSE(handle_->is_active());
+  EXPECT_FALSE(handle_->is_executing());
+
+  {
+    auto mock_get_status = mocking_utils::patch_and_return(
+      "lib:rclcpp_action", rcl_action_goal_handle_get_status, RCL_RET_ERROR);
+    EXPECT_FALSE(handle_->cancel());
```

Do this sets some error at some point?

---

I mean, the test case implies that it fails silently, which I don't think might be desirable

---

The coverage this test is trying to add is to the protected `try_canceling` method in `ServerGoalHandleBase`, which is a noexcept method, and just returns true or false whether cancel succeeded or failed. I exposed it publicly in `FibonacciServerGoalHandle` with `cancel()`, which does imply something slightly different with its name.

I changed `cancel()` to be named `try_cancel()` to give a better indication to its use.